include/aikido/control/detail/KinematicSimulationJointCommandExecutor-impl.hpp

```diff
@@ -0,0 +1,211 @@
+    if (mInProgress)
+    {
+      mInProgress = false;
+      mPromise->set_exception(
+          std::make_exception_ptr(std::runtime_error("Command canceled.")));
+    }
+    this->stop();
+  }
+}
+
+//==============================================================================
+template <ExecutorType T>
+std::future<int> KinematicSimulationJointCommandExecutor<T>::execute(
+    const std::vector<double> command,
+    const std::chrono::duration<double>& timeout)
+{
+
+  {
+    std::lock_guard<std::mutex> lock(mMutex);
+    DART_UNUSED(lock); // Suppress unused variable warning
+    auto promise = std::promise<int>();
+
+    if (command.size() != this->mJoints.size())
+    {
+      promise.set_exception(std::make_exception_ptr(
+          std::runtime_error("DOF of command does not match DOF of joints.")));
+      return promise.get_future();
+    }
+
+    if (mInProgress)
+    {
+      // Overwrite previous command
+      mCommand.clear();
+      mPromise->set_exception(
+          std::make_exception_ptr(std::runtime_error("Command canceled.")));
+    }
+
+    mPromise.reset(new std::promise<int>());
+
+    mCommand = command;
+    mInProgress = true;
+    mExecutionStartTime = std::chrono::system_clock::now();
+    mStartPosition.clear();
+    mTimeout = timeout;
+    switch (T)
+    {
+      case ExecutorType::VELOCITY:
+
+        for (size_t i = 0; i < mSkeleton->getDofs().size(); i++)
+        {
+          auto dof = mSkeleton->getDof(i);
+          dof->setVelocity(command[i]);
+          mStartPosition.push_back(dof->getPosition());
+        }
+
+        break;
+
+      case ExecutorType::POSITION:
+
+        for (size_t i = 0; i < mSkeleton->getDofs().size(); i++)
+        {
+          auto dof = mSkeleton->getDof(i);
+          mStartPosition.push_back(dof->getPosition());
+          if (mTimeout.count() > 1E-8)
+          {
+            dof->setVelocity(
+                (command[i] - mStartPosition[i]) / timeout.count());
+          }
+        }
+
+        break;
+
+      default:
+        // Other Executors not implemented
+        mCommand.clear();
+        mPromise->set_exception(std::make_exception_ptr(
+            std::logic_error("Executor not implemented")));
+        mInProgress = false;
+    }
+  }
+
+  return mPromise->get_future();
+}
+
+//==============================================================================
+template <ExecutorType T>
+void KinematicSimulationJointCommandExecutor<T>::step(
+    const std::chrono::system_clock::time_point& timepoint)
+{
+  std::lock_guard<std::mutex> lock(mMutex);
+
+  if (!mInProgress)
+    return;
+
+  const auto timeSinceBeginning = timepoint - mExecutionStartTime;
+  const auto executionTime
+      = std::chrono::duration<double>(timeSinceBeginning).count();
+
+  // executionTime may be negative if the thread calling \c step is queued
+  // before and dequeued after \c execute is called.
+  if (executionTime < 0)
+    return;
+
+  // Set joint states
+  double interpTime;
```

Comment:
Should this be inside the position case?

Reply:
Sure, just didn't do it due to compile error.

Fixed with scope addition: https://stackoverflow.com/questions/92396/why-cant-variables-be-declared-in-a-switch-statement